File: step_04_binary_search/4.2_bs_on_answers/P004_minimum_days_to_make_m_bouquets/solution.py
"""
Problem: Minimum Days to Make M Bouquets (LeetCode #1482)
Difficulty: MEDIUM | XP: 25

bloomDay[i] = day flower i blooms. Need m bouquets of k adjacent flowers.
Find minimum days to make m bouquets, or -1 if impossible.
"""
from typing import List


# ============================================================
# APPROACH 1: BRUTE FORCE (Try Every Unique Day)
# Time: O(n * max(bloomDay)) | Space: O(1)
# ============================================================
def brute_force(bloomDay: List[int], m: int, k: int) -> int:
    """Try each day from 1 to max(bloomDay). Return first day with enough bouquets."""
    n = len(bloomDay)
    if m * k > n:
        return -1

    def count_bouquets(day: int) -> int:
        bouquets = 0
        consecutive = 0
        for bd in bloomDay:
            if bd <= day:
                consecutive += 1
                if consecutive == k:
                    bouquets += 1
                    consecutive = 0
            else:
                consecutive = 0
        return bouquets

    for day in range(1, max(bloomDay) + 1):
        if count_bouquets(day) >= m:
            return day
    return -1


# ============================================================
# APPROACH 2: OPTIMAL (Binary Search on Days)
# Time: O(n * log(max(bloomDay) - min(bloomDay))) | Space: O(1)
# ============================================================
def optimal(bloomDay: List[int], m: int, k: int) -> int:
    """Binary search on the day in [min, max]. Check feasibility at each mid."""
    n = len(bloomDay)
    if m * k > n:
        return -1

    def count_bouquets(day: int) -> int:
        bouquets = 0
        consecutive = 0
        for bd in bloomDay:
            if bd <= day:
                consecutive += 1
                if consecutive == k:
                    bouquets += 1
                    consecutive = 0
            else:
                consecutive = 0
        return bouquets

    lo, hi = min(bloomDay), max(bloomDay)
    ans = -1
    while lo <= hi:
        mid = lo + (hi - lo) // 2
        if count_bouquets(mid) >= m:
            ans = mid
            hi = mid - 1    # feasible, try fewer days
        else:
            lo = mid + 1    # not enough bouquets, wait longer
    return ans


# ============================================================
# APPROACH 3: BEST (Binary Search + Early Termination)
# Time: O(n * log(max(bloomDay) - min(bloomDay))) | Space: O(1)
# ============================================================
def best(bloomDay: List[int], m: int, k: int) -> int:
    """Binary search with early termination in feasibility check."""
    n = len(bloomDay)
    if m * k > n:
        return -1

    def can_make(day: int) -> bool:
        bouquets = 0
        consecutive = 0
        for bd in bloomDay:
            if bd <= day:
                consecutive += 1
                if consecutive == k:
                    bouquets += 1
                    if bouquets >= m:    # early termination
                        return True
                    consecutive = 0
            else:
                consecutive = 0
        return False

    lo, hi = min(bloomDay), max(bloomDay)
    while lo < hi:
        mid = lo + (hi - lo) // 2
        if can_make(mid):
            hi = mid
        else:
            lo = mid + 1
    # Check if lo itself is feasible (handles edge case)
    return lo if can_make(lo) else -1


# ============================================================
# DRIVER
# ============================================================
if __name__ == "__main__":
    print("=== Minimum Days to Make M Bouquets ===\n")

    test_cases = [
        ([1, 10, 3, 10, 2], 3, 1, 3),
        ([1, 10, 3, 10, 2], 3, 2, -1),
        ([7, 7, 7, 7, 12, 7, 7], 2, 3, 12),
        ([1, 10, 2, 9, 3, 8, 4, 7, 5, 6], 4, 2, 9),
        ([1], 1, 1, 1),
    ]

    for bloomDay, m, k, expected in test_cases:
        b = brute_force(bloomDay, m, k)
        o = optimal(bloomDay, m, k)
        n = best(bloomDay, m, k)
        status = "PASS" if b == expected and o == expected and n == expected else "FAIL"
        print(f"Input: bloomDay={bloomDay}, m={m}, k={k}")
        print(f"  Brute:    {b}")
        print(f"  Optimal:  {o}")
        print(f"  Best:     {n}")
        print(f"  Expected: {expected}  [{status}]\n")

File: step_04_binary_search/4.2_bs_on_answers/P004_minimum_days_to_make_m_bouquets/Solution.java
/**
 * Problem: Minimum Days to Make M Bouquets (LeetCode #1482)
 * Difficulty: MEDIUM | XP: 25
 *
 * bloomDay[i] = day flower i blooms. Need m bouquets of k adjacent flowers.
 * Find minimum days to make m bouquets, or -1 if impossible.
 *
 * @author DSA_Nova
 */

// ============================================================
// Approach 1: Brute Force (Try Every Day)
// Time: O(n * max(bloomDay)) | Space: O(1)
// ============================================================
class BruteForce {
    public static int minDays(int[] bloomDay, int m, int k) {
        int n = bloomDay.length;
        if ((long) m * k > n) return -1;

        int maxDay = 0;
        for (int d : bloomDay) maxDay = Math.max(maxDay, d);

        for (int day = 1; day <= maxDay; day++) {
            if (countBouquets(bloomDay, day, k) >= m) return day;
        }
        return -1;
    }

    private static int countBouquets(int[] bloomDay, int day, int k) {
        int bouquets = 0, consecutive = 0;
        for (int bd : bloomDay) {
            if (bd <= day) {
                consecutive++;
                if (consecutive == k) {
                    bouquets++;
                    consecutive = 0;
                }
            } else {
                consecutive = 0;
            }
        }
        return bouquets;
    }
}

// ============================================================
// Approach 2: Optimal (Binary Search on Days)
// Time: O(n * log(max - min)) | Space: O(1)
// ============================================================
class Optimal {
    public static int minDays(int[] bloomDay, int m, int k) {
        int n = bloomDay.length;
        if ((long) m * k > n) return -1;

        int lo = Integer.MAX_VALUE, hi = Integer.MIN_VALUE;
        for (int d : bloomDay) {
            lo = Math.min(lo, d);
            hi = Math.max(hi, d);
        }

        int ans = -1;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (countBouquets(bloomDay, mid, k) >= m) {
                ans = mid;
                hi = mid - 1;
            } else {
                lo = mid + 1;
            }
        }
        return ans;
    }

    private static int countBouquets(int[] bloomDay, int day, int k) {
        int bouquets = 0, consecutive = 0;
        for (int bd : bloomDay) {
            if (bd <= day) {
                consecutive++;
                if (consecutive == k) {
                    bouquets++;
                    consecutive = 0;
                }
            } else {
                consecutive = 0;
            }
        }
        return bouquets;
    }
}

// ============================================================
// Approach 3: Best (Binary Search + Early Termination)
// Time: O(n * log(max - min)) | Space: O(1)
// ============================================================
class Best {
    public static int minDays(int[] bloomDay, int m, int k) {
        int n = bloomDay.length;
        if ((long) m * k > n) return -1;

        int lo = Integer.MAX_VALUE, hi = Integer.MIN_VALUE;
        for (int d : bloomDay) {
            lo = Math.min(lo, d);
            hi = Math.max(hi, d);
        }

        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (canMake(bloomDay, mid, m, k)) {
                hi = mid;
            } else {
                lo = mid + 1;
            }
        }
        return canMake(bloomDay, lo, m, k) ? lo : -1;
    }

    private static boolean canMake(int[] bloomDay, int day, int m, int k) {
        int bouquets = 0, consecutive = 0;
        for (int bd : bloomDay) {
            if (bd <= day) {
                consecutive++;
                if (consecutive == k) {
                    bouquets++;
                    if (bouquets >= m) return true; // early termination
                    consecutive = 0;
                }
            } else {
                consecutive = 0;
            }
        }
        return false;
    }
}

// Main driver
public class Solution {
    public static void main(String[] args) {
        System.out.println("=== Minimum Days to Make M Bouquets ===\n");

        int[][] bloomDays = {
            {1, 10, 3, 10, 2},
            {1, 10, 3, 10, 2},
            {7, 7, 7, 7, 12, 7, 7},
            {1, 10, 2, 9, 3, 8, 4, 7, 5, 6},
            {1}
        };
        int[] mArr = {3, 3, 2, 4, 1};
        int[] kArr = {1, 2, 3, 2, 1};
        int[] expected = {3, -1, 12, 9, 1};

        for (int t = 0; t < bloomDays.length; t++) {
            int b = BruteForce.minDays(bloomDays[t], mArr[t], kArr[t]);
            int o = Optimal.minDays(bloomDays[t], mArr[t], kArr[t]);
            int n = Best.minDays(bloomDays[t], mArr[t], kArr[t]);
            boolean pass = (b == expected[t] && o == expected[t] && n == expected[t]);

            System.out.println("Input:    bloomDay=" + java.util.Arrays.toString(bloomDays[t])
                + ", m=" + mArr[t] + ", k=" + kArr[t]);
            System.out.println("Brute:    " + b);
            System.out.println("Optimal:  " + o);
            System.out.println("Best:     " + n);
            System.out.println("Expected: " + expected[t]);
            System.out.println("Pass:     " + pass);
            System.out.println();
        }
    }
}
